### PR TITLE
Mccalluc/link user files to data sets

### PR DESCRIPTION
**Pull request by mccalluc**
On /user/files you can now click on the "document" icon to get get to /data_set.

### CODE REVIEW

File: refinery/analysis_manager/utils.py [hunk 1]
```diff
@@ -219,19 +218,7 @@ def fetch_objects_required_for_analysis(validated_analysis_config):
...
-            "Couldn't fetch Study from UUID: {} {}".format(study_uuid, e)
-        )
-
-    try:
-        data_set = InvestigationLink.objects.filter(
-            investigation__uuid=study.investigation.uuid
-        ).order_by("version").reverse()[0].data_set
-    except (AttributeError, IndexError) as e:
-        raise RuntimeError("Couldn't fetch DataSet {}".format(e))
+    data_set = get_data_set_for_study_uuid(study_uuid)
```
Comment (scottx611x) [inline refinery/analysis_manager/utils.py]: Maybe just `get_data_set_for_study` ?

Comment (mccalluc) [inline refinery/analysis_manager/utils.py]: The logic I needed to go from data_set to study already exists: Pull it out into a util function... but maybe it should be a method?

File: refinery/core/utils.py [hunk 2]
```diff
@@ -781,7 +781,8 @@ def delete_analysis_index(node_instance):
     """Remove a Analysis' related document from Solr's index.
     """
     try:
-        NodeIndex().remove_object(node_instance, using='data_set_manager')
+        data_set_manager.search_indexes.NodeIndex().remove_object(
+            node_instance, using='data_set_manager')
```
Comment (scottx611x) [inline refinery/core/utils.py]: Style: `)` on next line

File: refinery/core/utils.py [hunk 3]
```diff
@@ -993,3 +994,25 @@ def get_resources_for_user(user, resource_type):
...
+
+    try:
+        investigation_uuid = study.investigation.uuid
+        data_set = core.models.InvestigationLink.objects.filter(
+            investigation__uuid=investigation_uuid
+        ).order_by("version").reverse()[0].data_set
+    except (AttributeError, IndexError) as e:
+        raise RuntimeError(
+            "Couldn't fetch DataSet for Investigation {}: {}".format(
+                investigation_uuid, e)
```
Comment (scottx611x) [inline refinery/core/utils.py]: Style: `)` on next line

File: refinery/data_set_manager/search_indexes.py [hunk 4]
```diff
@@ -30,6 +31,7 @@ class NodeIndex(indexes.SearchIndex, indexes.Indexable):
     uuid = indexes.CharField(model_attr='uuid')
     study_uuid = indexes.CharField(model_attr='study__uuid')
     assay_uuid = indexes.CharField(model_attr='assay__uuid', null=True)
+    data_set_uuid = indexes.CharField(null=True)
```
Comment (mccalluc) [inline refinery/data_set_manager/search_indexes.py]: Note that there is no `model_attr`... Wondering if schema.xmls should be checked in, if they can be generated from this? There's a redundancy here.

File: refinery/data_set_manager/search_indexes.py [hunk 5]
```diff
@@ -53,14 +55,21 @@ def index_queryset(self, using=None):
...
+
+        try:
+            data_set = get_data_set_for_study_uuid(object.study.uuid)
+            data['data_set_uuid'] = data_set.uuid
+        except RuntimeError as e:
+            logger.warn(e)
 
         if object.assay is not None:
-            uuid += "_" + str(object.assay.id)
+            suffix += "_" + str(object.assay.id)
```
Comment (mccalluc) [inline refinery/data_set_manager/search_indexes.py]: Rename for clarity

File: refinery/data_set_manager/search_indexes.py [hunk 6]
```diff
@@ -53,14 +55,21 @@ def index_queryset(self, using=None):
...
         data = super(NodeIndex, self).prepare(object)
         annotations = AnnotatedNode.objects.filter(node=object)
-        uuid = str(object.study.id)
+        suffix = str(object.study.id)
+
+        try:
+            data_set = get_data_set_for_study_uuid(object.study.uuid)
+            data['data_set_uuid'] = data_set.uuid
+        except RuntimeError as e:
+            logger.warn(e)
```
Comment (mccalluc) [inline refinery/data_set_manager/search_indexes.py]: I hit this error: Thinking about it, I guess other node types don't have data_sets: Perhaps check the node type first, and only try to add data_set_uuid if there actually should be one?

File: refinery/data_set_manager/utils.py [hunk 7]
```diff
@@ -676,7 +676,7 @@ def generate_solr_params(params, assay_uuids, facets_from_config=False):
         solr_params += ''.join(
             [facet_template.format(s) for s
              in settings.USER_FILES_FACETS.split(",")])
-        solr_params += '&fl=*_generic_s,name,file_uuid,type,django_id'
+        solr_params += '&fl=*_generic_s,name,*_uuid,type,django_id'
```
Comment (mccalluc) [inline refinery/data_set_manager/utils.py]: We get a couple more fields back in the results which aren't used, but this feel less fragile.

File: refinery/solr/data_set_manager/conf/schema.xml [hunk 8]
```diff
@@ -156,6 +156,8 @@
 
     <field name="study_uuid" type="text_en" indexed="true" stored="true" multiValued="false" />
 
+    <field name="data_set_uuid" type="text_en" indexed="true" stored="true" multiValued="false" />
+
```
Comment (mccalluc) [inline refinery/solr/data_set_manager/conf/schema.xml]: see comment about redundancy above

File: refinery/solr/data_set_manager/conf/schema.xml [hunk 9]
```diff
@@ -156,6 +156,8 @@
 
     <field name="study_uuid" type="text_en" indexed="true" stored="true" multiValued="false" />
 
+    <field name="data_set_uuid" type="text_en" indexed="true" stored="true" multiValued="false" />
```
Comment (scottx611x) [inline refinery/solr/data_set_manager/conf/schema.xml]: Just take note that we'll need to run the `rebuild_index` command after this change: https://github.com/refinery-platform/refinery-platform/wiki/Solr-development#updated-index-model

File: refinery/data_set_manager/tests.py [hunk 10]
```diff
@@ -1637,7 +1637,14 @@ def test_get_basic_node(self):
 class NodeIndexTests(APITestCase):
 
     def setUp(self):
+        data_set = DataSet.objects.create()
```
Comment (scottx611x) [inline refinery/data_set_manager/tests.py]: maybe `create_dataset_with_necessary_models()` would be useful here?

File: refinery/analysis_manager/utils.py (hunk 1 above)
Comment (hackdna) [inline refinery/analysis_manager/utils.py]: In that case, it should probably be a method of the Study class with the following access pattern:
```python
study = Study.objects.get(uuid=study_uuid)
dataset = study.get_dataset()
```

File: refinery/data_set_manager/tests.py (hunk 10 above)
Comment (mccalluc) [inline refinery/data_set_manager/tests.py]: Maybe... though I'd need to then dig though the structure to get the objects I really need: This is a test about nodes down at the bottom, than datasets up at the top.

Maybe create_dataset_with_necessary_models could be changed to return a dict providing access to the internal objects? @scottx611x : Ok if I merge without this and maybe loop back later?

Comment (scottx611x) [inline refinery/data_set_manager/tests.py]: @mccalluc Sounds okay to me!

**Pull request merged**